extide/gradle/src/org/netbeans/modules/gradle/GradleConnectorManager.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.gradle;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.logging.Logger;
+import org.gradle.tooling.GradleConnector;
+import org.netbeans.api.project.Project;
+import org.openide.modules.OnStop;
+
+/**
+ *
+ * @author lkishalmi
+ */
+public final class GradleConnectorManager {
+    
+    private final Map<Project, GradleConnector> projectConnector = new WeakHashMap<>();
+    private final List<GradleConnector> connectors = new ArrayList<>();
+
+    private static final GradleConnectorManager INSTANCE = new GradleConnectorManager();
+    private static final Logger LOG = Logger.getLogger(GradleConnectorManager.class.getName());
+    
+    public static GradleConnectorManager getDefault() {
+        return INSTANCE;
+    } 
+    
+    public GradleConnector getConnector(Project prj) {
+        synchronized (connectors) {
+            GradleConnector ret = projectConnector.computeIfAbsent(prj, p -> {
+                GradleConnector conn = GradleConnector.newConnector();
+                connectors.add(conn);
+                return conn;
+            });
+            return ret;            
+        }
+    }
+    
+    
+    public void disconnectAll() {
+        LOG.info("Disconnecting from Gradle Daemons."); //NOI18N
+        synchronized (connectors) {
+            connectors.iterator().forEachRemaining(GradleConnector::disconnect);
```

Comment:
`connectors.forEach(GradleConnector::disconnect);` ?

Reply:
Well, it could be `forEach()` then a `clean()`. The plan is to have this method callable when needed, so it can free up resources, for example when the GRADLE_USER_HOME is changed in the IDE or for testing purposes. Not figured out the real need of the `disconnectAll()` invocations other than the IDE shutdown.

I think `GradleConnector` is a badly placed entrypoint to the Tooling API a strange mixture of a builder, some connection accounting, etc. 

Theoretically one instance of GradleConnector could be enough, though it is not thread safe and it's `useXXX`, `forXXX` methods are changing the state of an internal builder. That would require some kind of locking/synchronization. I hope I'd be wiser over time to deal with this better.

Reply:
not sure I understand. `forEachRemaining` isn't removing anything from the iterator.

You create a list iterator and iterate through it. This can be shortened into list.forEach(), both should be equivalent.

You can still clear the list afterwards of course.

Reply:
Wow! You are right! Probably I should not read JavaDoc after 4 hours sleep. I was so happy, that I found something new for draining a collection. It seems I've just mixed the `remove` and `forEachRemaining` while read that in the morning.

Thanks for putting me back to the track!

Reply:
no problem at all! this is a fairly exotic method tbh. I believe the reason why it is named that way is because a iterator can be at an arbitrary position, so forEach could be misleading if it is not a fresh iterator. Also get some sleep 4h is not enough :(